bpf: Inter-cluster SNAT with ClusterIP global service

YutaroHayakawa opened the pull request:
Implement very basic inter-cluster SNAT logic with ClusterIP global service. All the features are hidden behind the conditional macro, so it shouldn't affect the existing code path.

The datapath changes are split into small pieces. Here's a quick guide to how commits are organized.

- f1ed5fd600fba284495affd2dff3003fba7879b2 - 3b451b4251b496605a39d769449db9cac98c35f6 are preparatory commits (define macros, drop reason, etc...)
- 36f0789d5ee745a294c3aa3ea7513ba7e1461184 - c1f2237c6c649b89d31b8a46ba0be9627bc65624 are commits for actual datapath logic. The commits are ordered by packet flow, so if you read the code commit by commit, you'll automatically follow the packet flow. Labels in commit titles such as `Client/Egress` indicates which code path you're in.
- b5996c2268d6dfecf3ac62cf39fe4e7004591228 - 9a274dbe7f07a97b2a1749f067c75e91eecd4276 are tests.

```release-note
bpf: Inter-cluster SNAT with ClusterIP global service
```

YutaroHayakawa committed:
bpf: Add helper functions to get per-cluster CT

Add helper functions get_cluster_ct_map4/6 to get per-cluster conntrack
map. When cluster-aware addressing is disabled or cluster_id = 0, these
functions are identical to get_ct_map4/6.

We also added get_cluster_ct_any_map4/6 function that will be used to
get conntrack related map. Currently, &CT_MAP_ANY4/6 are hard-coded into
the code, but with cluster-aware addressing, we need to lookup
per-cluster conntrack map. Without cluster-aware addressing or
cluster_id = 0, it returns &CT_MAP_ANY4/6, so that it doesn't change
existing behavior.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Add helper functions to get per-cluster SNAT maps

Add helper functions get_cluster_snat_map_v4/6 to get per-cluster SNAT
maps. When cluster-aware addressing or inter-cluster SNAT are disabled
or cluster_id = 0, it just returns &SNAT_MAPPING_IPV4/6. So that
existing behavior won't be changed.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
Small bug fixes for per-cluster CT and SNAT map

- Define Map macro for per-cluster SNAT maps
- Fix wrong if condition in per-cluster CT map initialization

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Introduce from_tunnel field to ctmap value

Introduce a new field from_tunnel to ctmap value which indicates the
connection is established over tunnel. That means the reply traffic to
the connection always needs to go overlay even if the destination IP
address is reachable through underlay.

The use case is to avoid assymetric traffic when the packet is
inter-cluster SNAT'ed. In the current assumption of overlapping PodCIDR
support for Cluster Mesh, inter-cluster SNAT uses NodeIP as a source
address, thus it's reachable through underlay.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Ingress Reply path of the client cluster lxc

Transfer ClusterID using metadata from overlay, call into handle_policy
and lookup per-cluster conntrack map with ingress direction. It should
have the entry we created on the egress path. Thus, we do revDNAT for
service, bypass ingress policy and reach to the client Pod.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,test: Add BPF unit tests for inter-cluster SNAT communication

Added four new test files to test TCP three-way handshake with
inter-cluster SNAT communication.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,test: Add complexity-test scenarios for inter-cluster SNAT

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

brb commented on bpf/lib/lb.h:
```diff
@@ -1695,12 +1702,13 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
 	ctx_store_meta(ctx, CB_BACKEND_ID, state->backend_id);
 	ctx_store_meta(ctx, CB_CT_STATE, (__u32)state->rev_nat_index << 16 |
 		       state->loopback);
+	ctx_store_meta(ctx, CB_CLUSTER_ID, (__u32)cluster_id << 24);
```
Why do we need to do the left-shift? Can't we just cast to u32? Same argumentation applies to the stores above.

brb commented on bpf/bpf_lxc.c:
```diff
@@ -179,6 +187,22 @@ static __always_inline int __per_packet_lb_svc_xlate_6(void *ctx, struct ipv6hdr
+	if (dir == CT_EGRESS)
+		cluster_id = ctx_load_meta(ctx, CB_CLUSTER_ID) >> 24;
+	else if (dir == CT_INGRESS)
+		cluster_id = ctx_load_meta(ctx, CB_CLUSTER_ID_2) >> 24;
```
Maybe let's call it `CB_CLUSTER_ID_INGRESS`, and the former - `CB_CLUSTER_ID_EGRESS`?

YutaroHayakawa replied:
Ah, good idea. Let me fix that.

brb reviewed (approved):
Amazing PR, and thanks for the unit tests! Some minor non-blocking comment. It would be good to have a second look from someone from the core DP.

julianwiedmann commented on bpf/lib/l3.h:
```diff
@@ -144,8 +147,11 @@ static __always_inline int ipv6_local_delivery(struct __ctx_buff *ctx, int l3_of
+					       bool from_host __maybe_unused,
+					       bool hairpin_flow __maybe_unused,
+					       bool from_tunnel __maybe_unused,
+					       __u8 cluster_id __maybe_unused)
```
Note that all the `__maybe_unused` annotations are gone, and I think the `cluster_id` won't need it either.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -545,6 +546,8 @@ enum {
 #define DROP_NAT46		-187
 #define DROP_NAT64		-188
 #define DROP_POLICY_AUTH_REQUIRED	-189
+#define DROP_CT_NO_MAP_FOUND	-190
+#define DROP_SNAT_NO_MAP_FOUND	-191
```
These also need updates in `pkg/monitor/api/drop.go` and `api/v1/flow/flow.proto`

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -130,7 +130,8 @@ enum {
 #define CILIUM_CALL_IPV4_NODEPORT_DSR_INGRESS	40
 #define CILIUM_CALL_IPV6_NODEPORT_DSR_INGRESS	41
-#define CILIUM_CALL_SIZE			42
+#define CILIUM_CALL_IPV4_INTER_CLUSTER_REVSNAT	42
+#define CILIUM_CALL_SIZE			43
```
Need update in `test/bpf/check-complexity.sh`

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -205,6 +205,7 @@ static __always_inline __u8 __ct_lookup(const void *map, struct __ctx_buff *ctx,
 			ct_state->proxy_redirect = entry->proxy_redirect;
 			ct_state->from_l7lb = entry->from_l7lb;
 			ct_state->auth_required = entry->auth_required;
+			ct_state->from_tunnel = entry->from_tunnel;
```
This isn't relevant for `CT_SERVICE` lookups, right? Maybe you could turn that into an if-else statement with below check (same for `ct_create4()`).

With all the flags in ct_state, it's getting more difficult to follow where they actually apply... :)

YutaroHayakawa replied:
Yep, you're right. Fixed.

julianwiedmann reviewed (changes requested):
Added some minor comments.

For the full review, could you please try to split up `bpf,clustermesh: Basic inter-cluster SNAT communication with ClusterIP` into smaller chunks? Think the following could nicely come as preparatory patches:

- transporting the `from_tunnel` flag to bpf_lxc
- obtaining & transporting the `cluster_id` for a SVC backend, to pick the correct CT map

Not sure if the SNAT and revSNAT could be split into separate patches too ...

YutaroHayakawa committed:
bpf: Server/Ingress Request path of the inter-cluster communication

This commit implements the receive path of the cluster mesh with
overlapping PodCIDR on service backend side (not a revSNAT path).

When the packet comes from tunnel and go to local endpoints, propergate
information that the packet comes from tunnel to handle_policy call. It
finally reaches to the ct_create for the new connection and records
from_tunnel into conntrack entry. So that we can correctly put packets
back to tunnel on return path.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,test: Initialize per-cluster CT/SNAT for test

Initialize per-cluster CT/SNAT map when BPF_TEST macro is defined.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Egress Obtain service backend's ClusterID

Obtain service backend's ClusterID and transfer it to subsequent
tailcalls with lb4_ctx_store_state(). The rest of the datapath
identifes the destination (service backend) using IP + ClusterID.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Always enable per-packet LB for cluster-aware addressing

We must always enable per-packet LB when cluster-aware addressing is
enabled because SocketLB happens within Pod's namespace and we don't have
any good way to transfer out-of-band data (in our case, ClusterID of the
service backend) over network namespace boundary.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,hubble: Introduce DROP_CT_NO_MAP_FOUND

Introduce a new drop reason DROP_CT_NO_MAP_FOUND which can be used when
per-cluster conntrack lookup fails. Update cilium monitor and hubble
protobuf definitions appropreately.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Egress Lookup/Create per-cluster CT map entry on egress

Lookup per-cluster CT map on egress and use it. This only takes effect
when we have non-zero ClusterID obtained from service backend.
Otherwise, the behavior is the same as before.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Add helper functions to transfer ClusterID with mark

Introduce set_cluster_id_mark and ctx_get_cluster_id_from_mark which
set/get ClusterID to/from skb->mark. For XDP, they are nop. Also reserve
a new mark magic number for ClusterID.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Egress Cluster-aware egress network policy and tunnel redirect

Lookup ipcache with ClusterID. So that we can obtain correct security
identity and tunnel endpoint for the remote cluster endpoints with
overlapping PodCIDR.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,hubble: Introduce DROP_SNAT_NO_MAP_FOUND

Introduce a new drop reason DROP_SNAT_NO_MAP_FOUND which can be used when
per-cluster snat map lookup fails. Update cilium monitor and hubble
protobuf definitions appropreately.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Introduce CLUSTER_ID and IPV4_INTER_CLUSTER_SNAT macros

Introduce macros required for inter-cluster SNAT. CLUSTER_ID is a macro
that holds local ClusterID. It is used to make a decision whether we
should perform SNAT/RevSNAT for the packet on bpf_overlay.

IPV4_INTER_CLUSTER_SNAT is a source IPv4 address for inter-cluster SNAT.
It must be a NodeIP (which can be resolved to remote-node identity with
ipcache).

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Egress Inter-cluster SNAT egress

Introduce an egress side of the inter-cluster SNAT. The essential parts
are as follows.

1. We carry around the ClusterID with mark (for lxc -> overlay redirect)
   and skb->cb (for tailcalls).
2. We recognize the inter-cluster communication using the ClusterID. If
   the ClusterID is not 0 and not the same as local ClusterID, then it's
   inter-cluster communication.
3. To track SNAT mapping, we use per-cluster SNAT map.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Server/Egress Reply path of the inter-cluster communication

In the reply path of the inter-cluster communication, we lookup the
conntrack entry created on the request path. If the conntrack entry
has from_tunnel set and if the destination IP address is remote-node
IP (we can recognize it by the identity from ipcache), we set
tunnel_endpoint to destination IP address (remote-node IP, which should
also be a remote tunnel endpoint) and redirect packet to the tunnel
device.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Client/Ingress Inter-cluster SNAT ingress

Introduce an ingress side of the inter-cluster SNAT. First, we obtain
ClusterID from security identity and check if the packet is from the
remote cluster. After that, we check the destination IP address and if
the IP address is IPV4_INTER_CLUSTER_SNAT, we'll perform revSNAT.

We separate the revSNAT logic into tailcall since we hit the complexity
limit in the kernel 4.19.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Add a helper function to extract ClusterID from identity

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,test: Disable coverage reports for some tests

Disable coverage report for some tests affected by Coverbee's bug.

https://github.com/cilium/coverbee/issues/7

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1736,6 +1742,9 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
 	ctx_store_meta(ctx, CB_PROXY_MAGIC, 0);
+
+	*cluster_id = ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS) >> 24;
+	ctx_store_meta(ctx, CB_CLUSTER_ID_EGRESS, 0);
```
We could wrap the store/restore into `ENABLE_CLUSTER_AWARE_ADDRESSING` too, right?

YutaroHayakawa replied:
Thanks! Fixed!

julianwiedmann commented on bpf/lib/overloadable_skb.h:
```diff
@@ -67,6 +68,28 @@ set_encrypt_key_meta(struct __sk_buff *ctx, __u8 key, __u32 node_id)
+ * set_cluster_id_mark - sets the cluster_id mark.
+ */
+static __always_inline __maybe_unused void
+set_cluster_id_mark(struct __sk_buff *ctx, __u8 cluster_id)
```
For consistency, let's make this `ctx_set_cluster_id_mark()` ?

YutaroHayakawa replied:
Fixed!

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -557,7 +558,13 @@ int cil_to_overlay(struct __ctx_buff *ctx)
+	 * is a super set of the MARK_MAGIC_SNAT_DONE. They will never be used together,
+	 * but SNAT check should always take presedence.
+	 */
+	cluster_id = ctx_get_cluster_id_from_mark(ctx);
```
we could wrap this in some `ENABLE_` macros, right?

YutaroHayakawa replied:
Fixed!

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -597,6 +608,14 @@ static __always_inline bool snat_v4_prepare_state(struct __ctx_buff *ctx,
 		target->addr = IPV4_GATEWAY;
 		return true;
 	}
+
+# if defined(ENABLE_CLUSTER_AWARE_ADDRESSING) && defined(ENABLE_INTER_CLUSTER_SNAT)
+	if (target->cluster_id != 0 &&
+	    target->cluster_id != CLUSTER_ID) {
+		target->addr = IPV4_INTER_CLUSTER_SNAT;
+		return true;
+	}
+# endif
```
Is there a situation where this could collide with the earlier `IPV4_GATEWAY` check? :thinking:
Probably not, as we only care about pod-originating traffic.

But if yes, did you consider which check should take precedence (ie. which address should we use for SNATing such a connection?

YutaroHayakawa replied:
Yeah, so far, we only focus on the Pod-originated traffic, so it shouldn't be a problem. However, once we start supporting host-originated traffic, things may be changed. Let me leave this as is for now and revisit it when we support host-originated traffic.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2461,20 +2480,24 @@ lb_handle_health(struct __ctx_buff *ctx __maybe_unused)
 	__u16 proto;
 
+	ctx_store_meta(ctx, CB_CLUSTER_ID_EGRESS, (__u32)cluster_id << 24);
+
```
Just picking one line to comment on ... I'm not quite getting the shifting back&forth, but you probably need the remaining space for some other information?

Either way, it might help to have some store/load helpers that just do the shifting internally.

Working with the `__u8` vs `__u32` `cluster_id` feels like an over-complication ... is there a drawback to treating it as __u32 in all the generic paths, and treating the `__u8` range limitation just as a detail of the current implementation?

YutaroHayakawa replied:
Yeah, I agree with you. Treating cluster_id as `__u8` everywhere is a pain. However, `__u32` `cluster_id` doesn't fit into the mark (including magic number), it consumes more `skb->cb` bandwidth, and every time we downcast it to `__u8` (e.g., when we lookup map), we need to have range check. So, I ended up using `__u8`.

On the other hand, when we want to expand the cluster_id bit width at some point, This may be a pain since we need to expand the bit...

Ok, let me fix them to use `__u32` and do bounds check every time we downcast.

YutaroHayakawa replied:
Did above!

julianwiedmann reviewed (approved):
Thanks @YutaroHayakawa , awesome work! Think I get the picture of the different flows now, and the patch descriptions helped a lot too :pray:.

Some comments inline, but all just things that can be addressed as follow-on.

nathanjsweet commented on api/v1/flow/flow.proto:
```diff
@@ -375,6 +375,8 @@ enum DropReason {
     NAT46 = 187;
     NAT64 = 188;
     AUTH_REQUIRED = 189;
+    CT_NO_MAP_FOUND = 190;
```
@YutaroHayakawa FYI, @ysksuzuki is using the drop number 190 [in his PR](https://github.com/cilium/cilium/pull/23890/files#diff-74cced194f034ecadb39f852432c1822b6ef022065df0cb0d2ff6c718b339d45R378). Whoever is ready to merge first wins. Let me know when you're ready and I'll give you :green_apple: for API.

nbusseneau reviewed (approved):
`ci-structure` and `github-sec` were pinged for trivial changes, auto-acking.

YutaroHayakawa committed:
bpf: Define UINT8_MAX

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf: Change ipcache_lookup4/6 to accept __u32 cluster_id

Take cluster_id parameter as __u32 and convert it to __u8 internally
with overflow check. So that we can easily change the bit width of the
cluster_id for ipcache later.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bpf,hubble: Introduce DROP_INVALID_CLUSTER_ID

Introduce a new drop reason DROP_INVALID_CLUSTER_ID which can be used when
we see an invalid ClusterID (e.g. ClusterID doesn't fit into map key).
Update cilium monitor and hubble protobuf definitions appropreately.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

kaworu reviewed (approved):
Hubble protobuf changes LGTM, will let @nathanjsweet decide on the `DropReason` enum value race between this PR and https://github.com/cilium/cilium/pull/23890.

nathanjsweet reviewed (approved):
LGTM for API changes.

pchaigno commented on bpf/complexity-tests/netnext/bpf_lxc.txt:
```diff
@@ -1,3 +1,3 @@
--DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_ROUTING=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_SRV6=1 -DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_SESSION_AFFINITY=1 -DENAB ...
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_ROUTING=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_SRV6=1 -DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_SESSION_A ...
+-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_ROUTING=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_SRV6=1 -DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_SESSION_AFFINITY=1 -DENAB ...
 -DSKIP_DEBUG=1 -DENABLE_IPV6=1 -DENABLE_ROUTING=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DENABLE_ICMP_RULE=1 -DENABLE_CUSTOM_CALLS=1 -DENABLE_SRV6=1 -DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_SESSION_AFFINITY=1 -DENAB ...
```
@YutaroHayakawa Was `-DENABLE_CLUSTER_AWARE_ADDRESSING=1` not added to the IPv6-only case because it's not compatible with IPv6-only settings?

YutaroHayakawa replied:
IPv6 cluster-aware addressing is not yet supported.

aanm commented on bpf/lib/common.h:
```diff
@@ -850,7 +861,8 @@ struct ct_entry {
 	      from_l7lb:1, /* Connection is originated from an L7 LB proxy */
 	      auth_required:1,
-	      reserved:6;
+	      from_tunnel:1, /* Connection is over tunnel */
```
@YutaroHayakawa I noticed that we are missing some visibility information when we are dumping the CT map. Shouldn't we add a "FromTunnel" in [here](https://github.com/cilium/cilium/blob/1f7ed72e6b132202caaa085d6774bb62f58ca7b4/pkg/maps/ctmap/types.go#L555-L594)?
